common/serde-reflection/src/lib.rs

```diff
@@ -5,13 +5,110 @@
 
 //! # Serde Reflection (experimental)
 //!
-//! This crate provides a way to extract IDL-like format descriptions for values
-//! that implement the Serialize trait of Serde.
+//! This crate provides a way to extract IDL-like format descriptions for Rust containers
+//! that implement the Serialize and/or Deserialize trait(s) of Serde.
+//!
+//! ```rust
+//! # use libra_serde_reflection::*;
+//! # use serde::{Deserialize, Serialize};
+//! # #[derive(Serialize, PartialEq, Eq, Debug, Clone)]
+//! struct Name(String);
+//! #
+//! # impl<'de> Deserialize<'de> for Name {
+//! #     fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+//! #     where
+//! #         D: ::serde::Deserializer<'de>,
+//! #     {
+//! #         // Make sure to wrap our value in a container with the same name
+//! #         // as the original type.
+//! #         # [derive(Deserialize)]
+//! #         # [serde(rename = "Name")]
+//! #         struct InternalValue(String);
+//! #
+//! #         let value = InternalValue::deserialize(deserializer)?.0;
+//! #         // Enforce some custom invariant
+//! #         if value.len() >= 2 && value.chars().all(char::is_alphabetic) {
+//! #             Ok(Name(value))
+//! #         } else {
+//! #             Err(<D::Error as ::serde::de::Error>::custom(format!(
+//! #                 "Invalid name {}",
+//! #                 value
+//! #             )))
+//! #         }
+//! #     }
+//! # }
+//!
+//! # #[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+//! enum Person {
+//!     NickName(Name),
+//!     FullName { first: Name, last: Name },
+//! }
+//!
+//! # fn main() -> Result<(), Error> {
+//! // Start a session to trace formats.
+//! let mut tracer = Tracer::new(/* is_human_readable */ false);
+//!
+//! // For every type (here `Name`), if a user-defined implementation of `Deserialize`
+//! // exists that performs custom validation checks, start by tracing the serialization
+//! // of a valid value of this type.
+//! let bob = Name("Bob".into());
+//! tracer.trace_value(&bob)?;
+//!
+//! // Then, we may trace deserialization for top-level types (here `Person`) and their
+//! // dependencies as follows. This phase will use Rust values previously recorded
+//! // during serialization in order to pass custom checks.
+//! // TODO: Currently, recorded values are only used for `NewTypeStruct`s.
+//! let format = tracer.trace_type::<Person>()?;
+//! assert_eq!(format, Format::TypeName("Person".into()));
+//!
+//! // We may also request a sample value of type `Person`.
+//! let (person, _) = tracer.sample_type_once::<Person>()?;
+//! assert_eq!(person, Person::NickName(bob.clone()));
+//!
+//! // Stop the tracing session and obtain a final registry of containers (i.e. structs and enums).
+//! let registry = tracer.registry()?;
+//!
+//! // We have successfully extracted a format description of all Serde containers under `Person`.
```

Comment:
Serialization tracing is rather predictable: we traverse the user-provided value(s) as deep as we can and record everything useful on the way.

Deserialization tracing is a bit more tricky for two reasons:
* (coverage) We can only pick one variant case to explore each time we visit an enum. (We cannot "branch" because the provided visitors do not implement `Clone`) On the positive side, we are provided with the list of all variants, so it is easy to track which ones are left to be explored.
* (termination) In case of inductive data-structures (possibly mutually recursive), we need to stop the recursive exploration at some point to terminate.

Regarding variant coverage, the current approach is to have a loop outside of the graph exploration algorithm. For each iteration of the loop we should find one new variant of the traced type T, then stops.

Regarding termination, intuitively, we use a classical graph exploration algorithm and some notion of marking to avoid re-visiting the same nodes. However, when we wish to stop, the type system requires us to return a valid Rust value for V::Value, so this limits our choices to four stopping scenarios:
* while visiting an Option<T> we choose the value None;
* while visiting an Seq<T> we choose the empty sequence;
* while visiting an enum, we choose the first variant (ok, it is not always stopping the recursion, but close enough)
* while visiting a container (one of the 5 types), if the container's name is mapped to a recorded value, we MAY decide to use it. Currently, we always pick the recorded value for a `NewTypeStruct` and never do in the other cases.